### PR TITLE
Fix docs generation for cross-db sources in REDSHIFT RA3 node

### LINKED ISSUE #3236: cross db queries in redshift
### Describe the bug
When using source database that is different from the target (https://docs.getdbt.com/reference/resource-properties/database), I ran into an issue that adapter doesn't fetch columns, since it seems like it looks only into target db for meta information about them: https://github.com/fishtown-analytics/dbt/blob/develop/plugins/redshift/dbt/include/redshift/macros/adapters.sql#L79

### Steps To Reproduce
1. Have target db different from source db
2. Run model based on source table that doesn't exist on target using adapter macroses

### Expected behavior
Adapter expected to get meta information about source tables from source database 

### Screenshots and log output
Not applicable

### System information
**Which database are you using dbt with?**
- [ ] postgres
- [x] redshift
- [ ] bigquery
- [ ] snowflake
- [ ] other (specify: ____________)

**The output of `dbt --version`:**
```
0.17.2
```

**The operating system you're using:**
OSX

**The output of `python --version`:**
3.9.4

### Additional context
Not applicable

### PR DESCRIPTION
**Pull request by kostek-pl**
resolves #3236(partially)

This PR is related to #3236. It fixes documentation generation for cross-db objects used as SOURCES in models. However it not resolves full cross-db support as it's not possible at this moment due to AWS cross-db restrictions.

### Checklist
 - [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [x] I have run this code in development and it appears to resolve the stated issue
 - [x] This PR includes tests, or tests are not required/relevant for this PR
 - [x] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

### CODE REVIEW

File: plugins/postgres/dbt/adapters/postgres/impl.py [hunk 1]
```diff
@@ -112,10 +112,10 @@ def _link_cached_database_relations(self, schemas: Set[str]):
                 self.cache.add_link(referenced, dependent)
 
     def _get_catalog_schemas(self, manifest):
-        # postgres/redshift only allow one database (the main one)
+        # postgres/redshift(not ra3) only allow one database (the main one)
         schemas = super()._get_catalog_schemas(manifest)
         try:
-            return schemas.flatten()
+            return schemas.flatten(allow_duplicates=self.config.credentials.ra3_node)
```
Comment (jtcohen6) [inline plugins/postgres/dbt/adapters/postgres/impl.py:118]: Postgres credentials don't have an `ra3_node` property, so this is raising an error and breaking tests. I think the right move here is to reimplement `_get_catalog_schemas` within [`redshift/impl.py`](https://github.com/fishtown-analytics/dbt/blob/develop/plugins/redshift/dbt/adapters/redshift/impl.py), including this RA3-specific logic.

File: core/dbt/adapters/base/relation.py [hunk 2]
```diff
@@ -433,13 +433,14 @@ def search(
             for schema in schemas:
                 yield information_schema_name, schema
 
-    def flatten(self):
+    def flatten(self, allow_duplicates: bool=False):
```
Comment (jtcohen6) [inline core/dbt/adapters/base/relation.py:436]: Just a naming nit: this change isn't so much about duplicates per se; it's about allowing multiple databases at all. In the past, the only reason we'd expect to see multiple databases on Postgres/Redshift was if a metadata query accidentally returned duplicates.

**Review by jtcohen6 — COMMENTED** [review]
Thanks for the contribution @kostek-pl! I'm in favor of the overall concept. There's a little bit of reorganization needed in order to get the tests passing.

**Comment by kostek-pl** [comment]
Thank you very much @jtcohen6 for the review and tips.
I've updated code according to your suggestions, hope that it's done in the way you expected.

**Comment by jtcohen6** [comment]
Thanks @kostek-pl! The change here is looking good to me. Since we don't have an RA3 cluster available for automated tests, could you confirm that it resolves the issue for you locally? Namely, that you're able to generate docs that include sources from another database.

[A number of tests](https://app.circleci.com/pipelines/github/kostek-pl/dbt/4/workflows/e1fc2256-db6e-4f59-84a9-93befb7e8df1) were failing because they were missing secrets, on account of running from the `kostek-pl` CircleCI org instead of the `fishtown-analytics` CI org. I [re-ran in our org](https://app.circleci.com/pipelines/github/fishtown-analytics/dbt/2697/workflows/3ba03759-6cb8-4cc6-960b-576810484f90/jobs/42087), everything is passing except the linting step, which is just a handful of pep8 errors:
```
core/dbt/adapters/base/relation.py:436:53: E252 missing whitespace around parameter equals
core/dbt/adapters/base/relation.py:436:54: E252 missing whitespace around parameter equals
plugins/redshift/dbt/adapters/redshift/impl.py:70:1: W293 blank line contains whitespace
plugins/redshift/dbt/adapters/redshift/impl.py:73:16: E121 continuation line under-indented for hanging indent
plugins/redshift/dbt/adapters/redshift/impl.py:89:14: W291 trailing whitespace
```

- Could you remove this repo from your CircleCI org, so that it will run in our org instead?
- It looks like some of your commits have been co-authored by `kostkaw <wojciech.kostka@contractors.roche.com>`, which which hasn't signed the CLA. Could you amend those commits to be singularly authored by the email associated with your @kostek-pl account?

**Comment by kostek-pl** [comment]
@jtcohen6 thank you a lot for the remarks.
I've amended previous commits and got rid of my corporate email, removed CircleCi from my project and adjusted the code to be consistent with flake8.

Yes, I can confirm that This PR resolves the problem locally with docs generating.
To be more precise I've created 2 models, the first one reads(SOURCE) from the table that is created
in another database(sources) and the second one reads(REF) the first one which is on current database(dev).

with RA3_node not set or set to false I get an error:

Error sending message, disabling tracking
Compilation Error
  Cross-db references allowed only in redshift RA3.* node. Got {'sources', 'dev'}
  
when RA3_node is set to true then docs are generated

example used:

profile:
```yaml
[...]
target: dev
  outputs:
    dev:
      type: redshift
      ra3_node: true
[...]
```

schema:
```yaml
version: 2

sources:
  - name: sources
    database: sources
    schema: public
    tables:
      - name: cross_db_test_table

models:
    - name: cross_db_step1

    - name: cross_db_step2
```

cross_db_step1:
```sql
select * from {{ source('sources', 'cross_db_test_table') }}
```

cross_db_step2:
```sql
select * from {{ ref('cross_db_step1') }}
```

result:

image

**Review by jtcohen6 — COMMENTED** [review]
This looks great @kostek-pl!

Could you add a changelog entry, and add yourself to the list of contributors? We've just released 0.20.0rc1, and this PR represents net-new functionality, so it should really go under a new section `dbt next` (i.e. v0.21). If we end up putting out another release candidate for v0.20, however, I'll consider sneaking it in :)

**Review by jtcohen6 — APPROVED** [review]
Thanks for this @kostek-pl!

**Comment by jtcohen6** [comment]
> If we end up putting out another release candidate for v0.20, however, I'll consider sneaking it in :)

Update: We will be putting out v0.20.0rc2 to fix some bugs and installation issues with v0.20.0rc1. That said, we're going to keep the scope of v0.21 narrow, and be disciplined about putting out a first prerelease within 6-8 weeks of v0.20.0-final. So I'm going to stay disciplined here, too, and keep the change in v0.21.

**Pull request merged**